File: namegen.js

// Requires jQuery

function get_chunks(name) {
    var lead = /^[aeiouy]*(?:qu|[bcdfghjklmnpqrstvwxz])+/;
    var inner = /\B[aeiouy]+(?:qu|[bcdfghjklmnpqrstvwxz])+\B/g;
    var tail = /[aeiouy]+(?:qu|[bcdfghjklmnpqrstvwxz])*$/;
    name = name.toLowerCase();
    return [name.match(lead), name.match(inner), name.match(tail)];
}

function randint(min, max) {
    return Math.floor(Math.random() * (max - min + 1) + min);
}

function make_names(seed_names, amount=10, minimum_count=0, allow_short=true) {
    // Build our lists of usable chunks first.
    var lead_chunks = {}, inner_chunks = {}, tail_chunks = {};
    $.each(seed_names, function(index, name) {
        var chunk;
        var chunks = get_chunks(name);
        if (chunks[0]) {
            chunk = chunks[0][0];
            lead_chunks[chunk] = (lead_chunks[chunk] || 0) + 1;
        }
        if (chunks[1]) {
            $.each(chunks[1], function(index, chunk) {
                inner_chunks[chunk] = (inner_chunks[chunk] || 0) + 1;
            });
        }
        if (chunks[2]) {
            chunk = chunks[2][0];
            tail_chunks[chunk] = (tail_chunks[chunk] || 0) + 1;
        }
    });
    var leads = [], inners = [], tails = [];
    $.each(lead_chunks, function(chunk, count) {
        if (count >= minimum_count) {
            leads.push(chunk)
        }
    });
    $.each(inner_chunks, function(chunk, count) {
        if (count >= minimum_count) {
            inners.push(chunk)
        }
    });
    $.each(tail_chunks, function(chunk, count) {
        if (count >= minimum_count) {
            tails.push(chunk)
        }
    });
    // We've got our lists, now we can build us some names!
    var names = [];
    for (var i = 0; i < amount; i++) {
        var random_inners = randint(0, 1)
        random_inners += Number(!allow_short)
        if (Math.random() > 0.85) {
            random_inners += 1
        }
        var name = leads[randint(0, leads.length - 1)];
        for (var j = 0; j < random_inners; j++) {
            name += inners[randint(0, inners.length - 1)];
        }
        name += tails[randint(0, tails.length - 1)];
        name = name.charAt(0).toUpperCase() + name.slice(1);
        names.push(name);
    }
    return names;
}

File: namegen.py
import re
import random


def token_lists(names, cull=1,
                # Match 0 or more vowels + 1 or more consonants at the start of the word
                lead=r"^[aeiouy]*(?:qu|[bcdfghjklmnpqrstvwxz])+",
                # Match 1 or more vowels + 1 or more consonants inside a word (not start/end)
                inner=r"\B[aeiouy]+(?:qu|[bcdfghjklmnpqrstvwxz])+\B",
                # Match 1 or more vowels + 0 or more consonats at the end of a word
                trail=r"[aeiouy]+(?:qu|[bcdfghjklmnpqrstvwxz])*$"):
    lead = re.compile(lead)
    inner = re.compile(inner)
    trail = re.compile(trail)
    leads, inners, tails = {}, {}, {}
    # Populate dictionaries; key=pattern, value=frequency
    for name in names:
        name = name.lower()
        match = re.match(lead, name)
        if match:
            pat = match.group(0)
            count = leads.get(pat, 0)
            leads[pat] = count + 1
        matches = re.findall(inner, name)
        for pat in matches:
            count = inners.get(pat, 0)
            inners[pat] = count + 1
        match = re.search(trail, name)
        if match:
            pat = match.group(0)
            count = tails.get(pat, 0)
            tails[pat] = count + 1
    # Convert dicts to a list of tuples in the format (pattern, frequency)
    sort_key = lambda x: x[1]
    lead_srt = sorted(leads.items(), key=sort_key, reverse=True)
    inner_srt = sorted(inners.items(), key=sort_key, reverse=True)
    tail_srt = sorted(tails.items(), key=sort_key, reverse=True)
    # Build lists of patterns ordered most to least frequent and cull rares
    lead_list = [ x[0] for x in lead_srt if x[1] > cull ]
    inner_list = [ x[0] for x in inner_srt if x[1] > cull ]
    tail_list = [ x[0] for x in tail_srt if x[1] > cull ]
    return lead_list, inner_list, tail_list


def make_names(leads, inners, tails, amount=10):
    names = []
    for i in xrange(amount):
        syllables = random.randint(0,1)
        if random.random() > .85:
            syllables += 1
        name = random.choice(leads)
        for x in range(syllables):
            name += random.choice(inners)
        name += random.choice(tails)
        names.append(name.title())
    return names
